cpu/stm32f3/periph/i2c.c

```diff
@@ -0,0 +1,490 @@
+}
+
+int i2c_release(i2c_t dev)
+{
+    if (dev >= I2C_NUMOF) {
+        return -1;
+    }
+    mutex_unlock(&locks[dev]);
+    return 0;
+}
+
+int i2c_read_byte(i2c_t dev, uint8_t address, char *data)
+{
+    return i2c_read_bytes(dev, address, data, 1);
+}
+
+int i2c_read_bytes(i2c_t dev, uint8_t address, char *data, int length)
+{
+    I2C_TypeDef *i2c;
+
+    switch (dev) {
+#if I2C_0_EN
+        case I2C_0:
+            i2c = I2C_0_DEV;
+            break;
+#endif
+
+        default:
+            return -1;
+    }
+
+    /* start reception and send slave address */
+    _start(i2c, address, length, I2C_FLAG_READ);
+
+    /* read the data bytes */
+    _read(i2c, data, length);
+
+    /* end transmission */
+    _stop(i2c);
+
+    return length;
+}
+
+int i2c_read_reg(i2c_t dev, uint8_t address, uint8_t reg, char *data)
+{
+    return i2c_read_regs(dev, address, reg, data, 1);
+}
+
+int i2c_read_regs(i2c_t dev, uint8_t address, uint8_t reg, char *data, int length)
+{
+    I2C_TypeDef *i2c;
+
+    switch (dev) {
+#if I2C_0_EN
+        case I2C_0:
+            i2c = I2C_0_DEV;
+            break;
+#endif
+
+        default:
+            return -1;
+    }
+
+    /* send start sequence and slave address */
+    _start(i2c, address, length, I2C_FLAG_WRITE);
+
+    /* wait for ack */
+    DEBUG("Waiting for ACK\n");
+    while (!(i2c->ISR & I2C_ISR_TXIS));
+
+    /* send register number */
+    DEBUG("ACK received, write reg into DR\n");
+    i2c->TXDR = reg;
+
+    /* send repeated start sequence, read registers and end transmission */
+    DEBUG("ACK received, send repeated start sequence\n");
+    return i2c_read_bytes(dev, address, data, length);
+}
+
+int i2c_write_byte(i2c_t dev, uint8_t address, char data)
+{
+    return i2c_write_bytes(dev, address, &data, 1);
+}
+
+int i2c_write_bytes(i2c_t dev, uint8_t address, char *data, int length)
+{
+    I2C_TypeDef *i2c;
+
+    switch (dev) {
+#if I2C_0_EN
+        case I2C_0:
+            i2c = I2C_0_DEV;
+            break;
+#endif
+
+        default:
+            return -1;
+    }
+
+    /* start transmission and send slave address */
+    _start(i2c, address, length, I2C_FLAG_WRITE);
+
+    /* send out data bytes */
+    _write(i2c, data, length);
+
+    /* end transmission */
+    _stop(i2c);
+
+    return length;
+}
+
+int i2c_write_reg(i2c_t dev, uint8_t address, uint8_t reg, char data)
+{
+    return i2c_write_regs(dev, address, reg, &data, 1);
+}
+
+int i2c_write_regs(i2c_t dev, uint8_t address, uint8_t reg, char *data, int length)
+{
+    I2C_TypeDef *i2c;
+
+    switch (dev) {
+#if I2C_0_EN
+        case I2C_0:
+            i2c = I2C_0_DEV;
+            break;
+#endif
+
+        default:
+            return -1;
+    }
+
+    /* start transmission and send slave address */
+    /* increase length because our data is register+data */
+    _start(i2c, address, length+1, I2C_FLAG_WRITE);
+
+    /* wait for ack */
+    DEBUG("Waiting for ACK\n");
+    while (!(i2c->ISR & I2C_ISR_TXIS));
```

Comment:
@OlegHahm I understand your concerns. Any ideas what kind of break condition would fit here? Timer expiration?  We do have this situation in many places of peripheral drivers because you have to wait for the hardware to have finished a transaction, which is signalled with a flag, often in a status register or so..